cmd/govim/implements.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/govim/govim"
+	"github.com/govim/govim/cmd/govim/internal/golang_org_x_tools/lsp/protocol"
+	"github.com/govim/govim/cmd/govim/internal/golang_org_x_tools/span"
+	"github.com/govim/govim/cmd/govim/internal/types"
+)
+
+func (v *vimstate) implements(flags govim.CommandFlags, args ...string) error {
+	v.quickfixIsDiagnostics = false
+	b, pos, err := v.cursorPos()
+	if err != nil {
+		return fmt.Errorf("failed to get current position: %v", err)
+	}
+
+	params := &protocol.ImplementationParams{
+		TextDocumentPositionParams: protocol.TextDocumentPositionParams{
+			TextDocument: protocol.TextDocumentIdentifier{
+				URI: protocol.DocumentURI(b.URI()),
+			},
+			Position: pos.ToPosition(),
+		},
+	}
+
+	// must be non-nil
+	locs := []quickfixEntry{}
+
+	refs, err := v.server.Implementation(context.Background(), params)
+	if err != nil {
+		return fmt.Errorf("call to gopls.Implementation failed: %v", err)
+	}
+	if len(refs) == 0 {
+		return fmt.Errorf("unexpected zero length of implementations")
+	}
+	for _, ref := range refs {
```

Comment:
No problem at all! I've had a brief pass at factoring out the reuse, I think I will have to re-examine the sorting rather than just lifting it straight out of references as it does not reliably sort implements. I'll be able to give this a more detailed pass tomorrow.

Reply:
Please ignore my comment above: it does reliably sort implements into filenames, however I'm finding writing a test to prove this works over multiple files is a little trickier as gopls does not necessarily return the names in any order.

I've added a (crude?) check to distinguish where the first value needs to be discarded in the case of references.